### PR TITLE
warp torch tensors with homography in c++/cuda

### PR DESCRIPTION
**Pull request by balthazarneveu**
A simple Cuda Kernel to warp torch tensors using an homogrpahy

Cuda version
CPU version fallback (not optimized) Seems like the custom cuda version is much faster that the native torch combination of torch.nn.functional.affine_grid & torch.nn.functional.grid_sample.

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+## Test
+
+```
+pytest-3 test_warp.py
+```
+
+## Bench
+```
+python3 test_warp.py
```
Comment (balthazarneveu) [inline readme.md]:
```
python3 test_warp.py
```

Nvidia A100 80GB PCIe
```
Warp forward custom cuda kernel 0.0004 s, bandwidth 3158GB/s
Warp forward torch 0.0030 s, bandwidth 403GB/s
Speed up 7.7x
```
Nvidia RTX3090
```
Warp forward custom cuda kernel 0.0011 s, bandwidth 1065GB/s
Warp forward torch 0.0047 s, bandwidth 254GB/s
Speed up 4.1x
```

File: src/warp.cpp [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+               "Expecting the image and the homography to have the same batch size N.");
+    // check image shape and layout
+    AT_ASSERTM(image.dim() == NUM_DIMS, "Expecting a 4-dimensional image tensor");
+    AT_ASSERTM(homography.dim() == 3, "Expecting a 3-dimensional homography tensor");
+    AT_ASSERTM(homography.size(1) == 3, "Expecting a N,3,3 tensor");
+    AT_ASSERTM(homography.size(2) == 3, "Expecting a N,3,3 tensor");
+
+    torch::Tensor out_image;
+    if (width && height) {
+        out_image = torch::ones({image.size(BATCH_DIM), image.size(CHANNELS_DIM), height.value(), width.value()},
```
Comment (balthazarneveu) [inline src/warp.cpp]: torch.zeros  & torch.zeros_like⚠️

File: test_warp.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+def warp_perf(spatial_size):
+    n, ch = 16, 32  # batch and channels
+    data_size = 4
+    nbytes_read_write = data_size * (3 * 3 * n + 2 * ch * n
+                                     )  # read homography matrix + read and write (2*) one float per channel per batch
+    nbytes_read_write_torch = data_size * (2 * 3 * n + 2 * n + 2 * n * ch
+                                           )  # read affine + write grid + read and write warp
+    # NHWC tensor format for the inputs
+    x = torch.eye(spatial_size, dtype=torch.float32).repeat(n, ch, 1, 1).permute(0, 2, 3, 1).to(current_device)
```
Comment (balthazarneveu) [inline test_warp.py]: Channels last is not correct  here 🔴
